src/misc.c

```diff
@@ -581,21 +580,55 @@ char *human_prefix(uint64_t num)
 	return out;
 }
 
-char *human_prefix_small(double num)
+char *human_speed(uint64_t speed)
 {
 	char *out = mem_alloc_tiny(16, MEM_ALIGN_NONE);
-	char prefixes[] = "\0munp";
+	char prefixes[] = "\0KMGTPEZY";
 	char *p = prefixes;
 
-	while (p[1] && num > 0 && num < 1) {
-		num *= 1000;
+	while (p[1] && speed >= 1000000) {
+		speed /= 1000;
 		p++;
 	}
 
 	if (*p)
-		snprintf(out, 16, "%u %c", (uint32_t)num, *p);
+		snprintf(out, 16, "%u%c c/s", (uint32_t)speed, *p);
 	else
-		snprintf(out, 16, "%u ", (uint32_t)num);
+		snprintf(out, 16, "%u c/s", (uint32_t)speed);
+
+	return out;
+}
+
+char *human_prefix_small(double num)
+{
+	char *out = mem_alloc_tiny(16, MEM_ALIGN_NONE);
+	uint64_t number = num * 1E9;
+	int whole, milli, micro, nano;
+
+	nano = number % 1000;
+	number /= 1000;
+	micro = number % 1000;
+	number /= 1000;
+	milli = number % 1000;
+	whole = number / 1000;
+
+	if (whole) {
+		if (milli)
+			snprintf(out, 16, "%d.%03d ", whole, milli);
+		else
+			snprintf(out, 16, "%d ", whole);
+	} else if (milli) {
+		if (micro)
+			snprintf(out, 16, "%d.%03d m", milli, micro);
+		else
+			snprintf(out, 16, "%d m", milli);
+	} else if (micro) {
+		if (nano)
+			snprintf(out, 16, "%d.%03d u", micro, nano);
+		else
+			snprintf(out, 16, "%d u", micro);
```

Comment:
Yes... but it would have been more relevant before we moved it to a shared function: It was only used for OpenCL auto-tune before and the rationale was to save precious screen estate but in practice they always end up with three decimals anyway so it was moot. Better would have been to just drop a decimal or two (and have it proper rounded as opposed to truncated).

Now though, since we're using this function for `--list=build-info` I think that would mean we'd print resolution/latency figures as eg `1.000 us` instead of `1 us` and that would be misleading, wouldn't it? At the very least we'd still have to print `28 ns` (my actual latency for the hi-res timer) and definitely not `28.000 ns`, right?

Reply:
On one hand, you're right, '28.000 ns` would be weird for anything, and `1.000 us` would be weird when we had just `1` because of low precision elsewhere. On the other hand, the current code could also print e.g. `1.100 us` when those `00` do not necessarily reflect precision of the input to this function, so making just the case of `000` special is inconsistent.

As to the `28.000 ns` example again, the current code simply never prints decimals of `nano`, so this is already impossible, and that's fine.

So I once again suggest that we simply drop the special cases for omitting `000`. If you like to omit that when we _know_ the input had lower precision, that should be passed into the function out-of-band (or a separate function used), not inferred (sometimes wrongly) from the value being printed.

Reply:
It's just that (exaggerating a little) I'd rather always *drop* the decimals, but then we're back with the original problems of #4533 (lol). The original functions' goals wasn't precision, it was the opposite: Omit irrelevant precision and print a "simple number" for a human (hence the function names). The problem is obviously how to know just what precision is relevant or not and this gets harder the more we try to make them shared functions for use in varying context.

Anyways I'll probably end up doing just what you said :-)